From 29389209053bcced7552177e71383d9f5f1356a2 Mon Sep 17 00:00:00 2001
From: Chuanxiao Dong <chuanxiao.dong@intel.com>
Date: Sun, 10 Mar 2024 09:50:04 +0800
Subject: [PATCH 16/25] KVM: TDX: Use the same shared EPT for L1 and L2

The mapping in shared EPT is controlled by host VMM and is always
untrusted to all the L1 and L2 VMs. Not necessary to create dedicated
shared EPT for each L1 and L2 VMs, but simply use the same the same one.

Signed-off-by: Chuanxiao Dong <chuanxiao.dong@intel.com>
---
 arch/x86/kvm/vmx/tdx.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/arch/x86/kvm/vmx/tdx.c b/arch/x86/kvm/vmx/tdx.c
index 75053ba3bf2b..ddb385d48e75 100644
--- a/arch/x86/kvm/vmx/tdx.c
+++ b/arch/x86/kvm/vmx/tdx.c
@@ -1603,8 +1603,13 @@ static int handle_tdvmcall(struct kvm_vcpu *vcpu)
 
 void tdx_load_mmu_pgd(struct kvm_vcpu *vcpu, hpa_t root_hpa, int pgd_level)
 {
+	struct kvm_tdx *kvm_tdx = to_kvm_tdx(vcpu->kvm);
+	int i;
+
 	WARN_ON_ONCE(root_hpa & ~PAGE_MASK);
 	td_vmcs_write64(to_tdx(vcpu), SHARED_EPT_POINTER, root_hpa);
+	for (i = 0; i < kvm_tdx->num_l2_vms; i++)
+		l2td_vmcs_write64(to_tdx(vcpu), SHARED_EPT_POINTER, root_hpa, i);
 }
 
 static void tdx_unpin(struct kvm *kvm, kvm_pfn_t pfn, enum pg_level level)
-- 
2.34.1

